summoner-cli/src/Summoner/Tree.hs

```diff
@@ -23,10 +26,52 @@ data TreeFs
 -- | Walks through directory tree and write file contents, creating all
 -- intermediate directories.
 traverseTree :: TreeFs -> IO ()
+traverseTree (File name content) = writeFileText name content
 traverseTree (Dir name children) = do
     createDirectoryIfMissing False name
     withCurrentDirectory name $ for_ children traverseTree
-traverseTree (File name content) = writeFileText name content
+
+{- | This function converts a string file path to the tree structure.
+
+For a path like this: @".github/workflow/ci.yml"@
+
+This function produces the following tree:
+
+@
+.github/
+└── workflow/
+    └── ci.yml
+@
+-}
+pathToTree :: FilePath -> Text -> TreeFs
+pathToTree path content =
+    let pathParts = splitDirectories path
+    in case pathParts of
+        [] -> Dir path []  -- shouldn't happen
+        x:xs -> go x xs
+  where
+    go :: FilePath -> [FilePath] -> TreeFs
+    go p [] = File p content
+    go p (x:xs) = Dir p [go x xs]
+
+{- | This functions inserts given 'TreeFs' node into the list of existing
+'TreeFs' nodes. The behavior of this function is the following:
+
+1. It merges duplicating directories.
+2. It overrides existing 'File' with the given 'TreeFs' in case of duplicates.
+-}
+insertTree :: TreeFs -> [TreeFs] -> [TreeFs]
+insertTree node [] = [node]
+insertTree node (x:xs) = case (node, x) of
+    (Dir _ _, File _ _) -> x : insertTree node xs
+    (File _ _, Dir _ _) -> x : insertTree node xs
+    (File nodePath _, File curPath _)
+        | nodePath == curPath -> node : xs
+        | otherwise -> x : insertTree node xs
```

Comment:
I don't quite get this case. When this file is already in the tree, it inserts it, but if it's not, it somehow continues inserting on the next layer? How so?

Reply:
Right, the implementation is not obvious, but I tried to do my best. This function always inserts `node` to the list, at least when it reaches the empty list at the end:

```haskell
insertTree node [] = [node]
```

In this case, if this function finds two `File`s with the same path, it will replace current `x` with `node` that why it's `node : xs`. This is done if people want to override default files like `.travis.yml` or `Prelude.hs` with their files. If file paths are not equal, then we keep `x`, and we will try to insert `node` into `xs`. Eventually, we either reach the end of the list, and we will add `node` unconditionally, or we will find a `File` that has the same name and will replace it.

Reply:
Hmm, but are other nodes placed unconditionally? I thought there is some kind of hierarchy so no need to go till the end of the list each time.

Reply:
Well, this list is not sorted at this stage. It could be `[File "foo" x, File "bar" y]` or `[File "bar" x, File "foo" y]`. I checked the implementation, and it looks like we sort this list only when printing to terminal or in TUI.

Now, after you told this, I started to think that probably there's a better way to implement `TreeFs`. At least `TreeFs` shouldn't have duplicate filepaths but since we're using lists, it's possible to have a bug in the implementation and introduce duplicate fields.

Maybe instead of

```haskell
data TreeFs
    = Dir FilePath [TreeFs]
    | File FilePath Text
```

we should have

```haskell
data TreeFs
    = Dir (Map FilePath TreeFs)
    | File Text
```

The latter representation won't allow having duplicate filepaths. But I believe that the current algorithm is also correct, even if the logic is not straightforward... And changing this type is a big rewrite...

Reply:
Thank you for the clarification, Dmitrii! Yeah, I don't think that rewrite is the right step right now, and we could keep this function as it is for this PR. But let's think more about this sorting before option and create the issue if we will decide that it could work better and would become easier to maintain.

Reply:
I've created the issue:

* https://github.com/kowainik/summoner/issues/371